### PR TITLE
Ml model ensembles

### PR DESCRIPTION
**Pull request by hmacdope**
## Description
Adds ability to pull an ML model ensemble from a remote specification at a URL and also to do inference with the ML model ensemble. 

## Developers certificate of origin
- [X] I certify that this contribution is covered by the MIT license as defined in our [LICENSE](https://github.com/choderalab/asapdiscovery/blob/main/LICENSE) and adheres to the [**Developer Certificate of Origin**](https://developercertificate.org/).

### CODE REVIEW

**Comment by hmacdope** [comment]
@kaminow  I would love to hear your thoughts here.

**Review by kaminow — CHANGES_REQUESTED** [review]
a couple nitpicky things but overall looks really good!

File: asapdiscovery-ml/asapdiscovery/ml/models.py [hunk 1]
```diff
@@ -150,7 +158,212 @@ def pull(self, local_dir: Union[Path, str] = None) -> "LocalMLModelSpec":
...
+                        type=model_data["type"],
+                        base_url=model_data["base_url"],
+                        weights_resource=submodel[subname]["resource"],
+                        weights_sha256hash=submodel[subname]["sha256hash"],
+                        config_resource=model_data["config"]["resource"],
+                        config_sha256hash=model_data["config"]["sha256hash"],
+                        last_updated=(
+                            model_data["last_updated"]
+                            if "last_updated" in model_data
+                            else "2024-02-06"
```
[the remark this answers is not in the document]
Comment (hmacdope) [reply asapdiscovery-ml/asapdiscovery/ml/models.py:307]: Ah good catch, debugging.

File: asapdiscovery-ml/asapdiscovery/ml/models.py [hunk 2]
```diff
@@ -150,7 +158,212 @@ def pull(self, local_dir: Union[Path, str] = None) -> "LocalMLModelSpec":
...
+            )
+        if len({model.mtenn_upper_pin for model in models}) > 1:
+            raise ValueError(
+                "All models in an ensemble must have the same mtenn_upper_pin"
+            )
+        return models
+
+    @validator("models")
+    @classmethod
+    def check_all_last_updated(cls, models):
```
[the remark this answers is not in the document]
Comment (hmacdope) [reply asapdiscovery-ml/asapdiscovery/ml/models.py:195]: No I suppose not, will remove.

File: asapdiscovery-ml/asapdiscovery/ml/models.py [hunk 3]
```diff
@@ -317,31 +545,39 @@ def from_yaml(cls, yaml_file: Union[str, Path]) -> "MLModelRegistry":
...
-                mtenn_upper_pin=(
-                    model_data["mtenn_upper_pin"]
-                    if "mtenn_upper_pin" in model_data
-                    else None
-                ),
+            is_ensemble = (
+                "remote_ensemble" in model_data and model_data["remote_ensemble"]
             )
+            if is_ensemble:
+                models[model] = RemoteEnsembleHelper.to_ensemble_spec()
```
[the remark this answers is not in the document]
Comment (hmacdope) [reply asapdiscovery-ml/asapdiscovery/ml/models.py:552]: yes good catch

File: asapdiscovery-ml/asapdiscovery/ml/models.py [hunk 4]
```diff
@@ -150,7 +158,212 @@ def pull(self, local_dir: Union[Path, str] = None) -> "LocalMLModelSpec":
...
+            sha256hash: 0b2
+        - asapdiscovery-GAT-Z:
+            resource: asapdiscovery-GAT-Z.th
+            sha256hash: 0b3
+    config:
+        resource: asapdiscovery-SARS-CoV-2-Mpro.json
+    targets:
+        - SARS-CoV-2-Mpro
+        - MERS-CoV-Mpro
+    mtenn_lower_pin: "0.5.0"
```
[the remark this answers is not in the document]
Comment (hmacdope) [reply asapdiscovery-ml/asapdiscovery/ml/models.py:256]: I think lets just leave it as specified for the ensemble.

File: asapdiscovery-ml/asapdiscovery/ml/models.py [hunk 5]
```diff
@@ -150,7 +158,212 @@ def pull(self, local_dir: Union[Path, str] = None) -> "LocalMLModelSpec":
...
+        Check that all models in the ensemble are compatible with the same mtenn version
+        """
+        if len({model.mtenn_lower_pin for model in models}) > 1:
+            raise ValueError(
+                "All models in an ensemble must have the same mtenn_lower_pin"
+            )
+        if len({model.mtenn_upper_pin for model in models}) > 1:
+            raise ValueError(
+                "All models in an ensemble must have the same mtenn_upper_pin"
+            )
```
[the remark this answers is not in the document]
Comment (hmacdope) [reply asapdiscovery-ml/asapdiscovery/ml/models.py]: Done

**Comment by hmacdope** [comment]
@kaminow probably ready for another look.

File: asapdiscovery-ml/asapdiscovery/ml/models.py [hunk 6]
```diff
@@ -150,7 +158,253 @@ def pull(self, local_dir: Union[Path, str] = None) -> "LocalMLModelSpec":
...
+            m2_version_compat = low_pin2 <= cur_version < upper_pin2
+
+            if not m1_version_compat or not m2_version_compat:
+                return False
+
+            # now check m1 and m2 are cross compatible
+            if not (low_pin1 <= upper_pin2 and low_pin2 <= upper_pin1):
+                return False
+
+    return True
```
Comment (kaminow) [inline asapdiscovery-ml/asapdiscovery/ml/models.py:277]: this works so fine to leave it is as, but could also do something like gather all lower pins and upper pins, then just make sure the `mtenn` version is between the highest lower pin and lowest upper pin (I think this is equivalent)

File: asapdiscovery-ml/asapdiscovery/ml/models.py [hunk 7]
```diff
@@ -315,33 +584,47 @@ def from_yaml(cls, yaml_file: Union[str, Path]) -> "MLModelRegistry":
...
-                    model_data["mtenn_upper_pin"]
-                    if "mtenn_upper_pin" in model_data
-                    else None
-                ),
-            )
+            try:
+                model_data = spec[model]
+                has_config = "config" in model_data
+                is_ensemble = (
+                    "remote_ensemble" in model_data and model_data["remote_ensemble"]
```
Comment (kaminow) [inline asapdiscovery-ml/asapdiscovery/ml/models.py:591]: is this documented anywhere? sorry if I'm just being dumb and missing it

**Review by kaminow — APPROVED** [review]
two minor suggestions/questions but other than that looks good!

**Pull request merged**